typing/typecore.ml

```diff
@@ -3199,95 +3226,137 @@ let rec approx_type env sty =
       approx_type env sty
   | _ -> newvar ()
 
-let type_pattern_approx env spat =
+let type_pattern_approx env spat ty_expected =
   match spat.ppat_desc with
-  | Ppat_constraint (_, sty) -> approx_type env sty
-  | _ -> newvar ()
+  | Ppat_constraint (_, sty) ->
+      let inferred_ty = approx_type env sty in
+      begin try unify_pat_types spat.ppat_loc env inferred_ty ty_expected
+        with Unify trace ->
+        raise(Error(spat.ppat_loc, env, Pattern_type_clash(trace, None)))
+      end;
+  | _ -> ()
 
-let type_approx_fun env label default spat ret_ty =
-  let ty = type_pattern_approx env spat in
-  let ty =
-    match label, default with
-    | (Nolabel | Labelled _), _ -> ty
-    | Optional _, None ->
-       unify_pat_types spat.ppat_loc env ty (type_option (newvar ()));
-       ty
-    | Optional _, Some _ ->
-       type_option ty
+let type_approx_fun_one_param
+  env label default spato ty_expected ~first ~in_function =
+  let ty_arg, ty_ret =
+    try filter_arrow env ty_expected label
+    with Filter_arrow_failed err ->
+      let loc_fun, ty_fun = in_function in
+      let err =
+        error_of_filter_arrow_failure ~explanation:None ty_fun err ~first
+      in
+      raise (Error(loc_fun, env, err))
+  in
+  let () =
+    match spato with
+    | None -> ()
+    | Some spat ->
+        let ty_arg =
+        match label, default with
+        | (Nolabel | Labelled _), _ -> ty_arg
+        | Optional _, None ->
+          unify_pat_types spat.ppat_loc env ty_arg (type_option (newvar ()));
+          ty_arg
+        | Optional _, Some _ ->
+          let var = newvar () in
+          unify_pat_types spat.ppat_loc env ty_arg (type_option var);
+          var
+        in
+        type_pattern_approx env spat ty_arg;
   in
-  newty (Tarrow (label, ty, ret_ty, commu_ok))
+  ty_ret
 
-let type_approx_constraint env ty constraint_ ~loc =
+let type_approx_constraint env constraint_ ~loc ty_expected =
   match constraint_ with
   | Pconstraint constrain ->
       let ty_constrain = approx_type env constrain in
-      begin try unify env ty ty_constrain with Unify err ->
+      begin try unify env ty_constrain ty_expected with Unify err ->
         raise (Error (loc, env, Expr_type_clash (err, None, None)))
       end;
       ty_constrain
-  | Pcoerce (constrain, coerce) ->
-      let approx_ty_opt = function
-        | None -> newvar ()
-        | Some sty -> approx_type env sty
-      in
-      let ty_constrain = approx_ty_opt constrain
-      and ty_coerce = approx_type env coerce in
-      begin try unify env ty ty_constrain with Unify err ->
+  | Pcoerce (_constrain, coerce) ->
+      let ty_coerce = approx_type env coerce in
+      begin try unify env ty_coerce ty_expected with Unify err ->
         raise (Error (loc, env, Expr_type_clash (err, None, None)))
       end;
-      ty_coerce
+      ty_expected
```

Comment:
This type here is formally wrong. It is not the expected type of the expression inside the constraint. However as this type is never used (see `ignore` at call site) it is not problematic. But this type still looks correct which could lead to potential issue in the future. I propose to replace `ty_expected` here by a trivially wrong type such as `Tuple []` to prevent people from using it.

Another approach would be to refactor this function into only one case : taking a `approx_type_with_expected`

Reply:
This type was used later on an causes a problem in the following example :
```ocaml
module M : sig
   type t = private int
   val x : t
end = struct
   type t = int
   let x = 0
end

let rec f () : M.t :> int = (M.x : M.t)
```

We need to either return the approximation `type_approx constrain` or `newvar ()`